test/dowloadService.test.ts

```diff
@@ -0,0 +1,83 @@
+import { describe, test, afterEach, vi, expect } from 'vitest';
+import { DownloadService } from '../src/components/downloadService';
+import type { FileSystem } from 'vscode';
+import { AxiosError, AxiosInstance } from 'axios';
+import nock from 'nock';
+import axiosInstance from '../src/axios';
+
+const mockedFileSystemUtilities = {
+	getModificationTime: vi.fn(() => 1),
+	setChmod: vi.fn(),
+};
+
+const fs = {
+	writeFile: vi.fn(),
+	stat: vi.fn(() => ({ mtime: Date.now() })),
+} as unknown as FileSystem;
+
+const downloadService = new DownloadService(
+	fs,
+	// @ts-ignore
+	mockedFileSystemUtilities,
+);
+
+const NETWORK_ERROR = new AxiosError('Some connection error');
+NETWORK_ERROR.code = 'ECONNRESET';
+
+// 3 failed responses, then good response
+const responses = [
+	() => nock('https://test.com').head('/test').replyWithError(NETWORK_ERROR),
+	() => nock('https://test.com').head('/test').replyWithError(NETWORK_ERROR),
+	() => nock('https://test.com').head('/test').replyWithError(NETWORK_ERROR),
+	() =>
+		nock('https://test.com')
+			.head('/test')
+			.reply(200, '', { 'last-modified': new Date(2).toISOString() }),
+	() => nock('https://test.com').get('/test').replyWithError(NETWORK_ERROR),
+	() => nock('https://test.com').get('/test').replyWithError(NETWORK_ERROR),
+	() => nock('https://test.com').get('/test').replyWithError(NETWORK_ERROR),
+	() => nock('https://test.com').get('/test').reply(200, 'Test'),
+];
+
+const setupResponses = (client: AxiosInstance, responses: Array<Function>) => {
+	const configureResponse = () => {
+		const response = responses.shift();
+		if (response) {
+			response();
+		}
+	};
+
+	client.interceptors.request.use(
+		(config) => {
+			configureResponse();
+			return config;
+		},
+		(error) => {
+			configureResponse();
+			return Promise.reject(error);
+		},
+	);
+};
+
+describe('DownloadService', () => {
+	afterEach(() => {
+		nock.cleanAll();
+		nock.enableNetConnect();
+	});
+
+	test('Should retry 3 times if request fails', async () => {
+		setupResponses(axiosInstance, responses);
+
+		await downloadService.downloadFileIfNeeded(
+			`https://test.com/test`,
+			// @ts-expect-error passing a string instead of URI, because URI cannot be imported from vscode
+			'/',
+			'755',
+		);
+
+		expect(fs.writeFile).toBeCalledWith(
+			'/',
+			new Uint8Array([84, 101, 115, 116]),
```

Comment:
what does that mean?

Reply:
We mock `get` response with `Test`, and expect that writeFile will be called with uint representation of that string. fs.writeFile is mocked, so we can check args it was called with

Reply:
Ah ok, the array in question is basically `Test`:
```
new Uint8Array([84, 101, 115, 116]).forEach((x) => console.log(String.fromCharCode(x)))
```